### DISCUSSION TITLE
GraphQL api error

### DISCUSSION THREAD

**Discussion by chilyanin** [community/community#39332] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I'm using python with requests and getting error in mutation...Help pls 
	current_repo_permission = current_repo['permission']
	current_repo_id = current_repo['node']['id']
	graphql_mutation = """mutation ($repositoryId:String!, $teamIds:String!, $permission:String!) {
      updateTeamsRepository(
        input: {repositoryId: $repositoryId, teamIds: $teamIds, permission: $permission}
      )
	}"""
	variables = {"repositoryId": current_repo_id, "teamIds": team_dst_id, "permission": current_repo_permission}
	response = requests.post(url, headers=headers, json={'query': graphql_mutation, 'variables': variables}) 

Error: {"errors":[{"path":["mutation","updateTeamsRepository"],"extensions":{"code":"selectionMismatch","nodeName":"field \'updateTeamsRepository\'","typeName":"UpdateTeamsRepositoryPayload"},"locations":[{"line":2,"column":7}],"message":"Field must have selections (field \'updateTeamsRepository\' returns UpdateTeamsRepositoryPayload but has no selections. Did you mean \'updateTeamsRepository { ... }\'?)"}

**Comment by LangLangBart** [comment, marked as answer]
adding `clientMutationId`  works ?

```graphql
mutation ($repositoryId: String!, $teamIds: String!, $permission: String!) {
  updateTeamsRepository(input: {repositoryId: $repositoryId, teamIds: $teamIds, permission: $permission}) {
    clientMutationId
  }
}
```

**Reply by chilyanin** [reply]
Oh wonderful !!! Thank you very much. What was that ? Have some type mismatch errors but after correcting it works

**Comment by chilyanin** [comment]
Should i always add Return fields to my mutations ?

**Reply by LangLangBart** [reply]
I assume that GraphQL must always have at least one return field for a mutation, even if no exclamation mark indicates the need.

### Example - this will not work

```graphql
mutation {
  addReaction(input: {subjectId: "I_kwDOIV8V6M5WM_mv", content: HOORAY})
}
```

### Example - this will work
```graphql
mutation {
  addReaction(input: {subjectId: "I_kwDOIV8V6M5WM_mv", content: HOORAY}) {
    clientMutationId
  }
}
```

![SCR-20221117-pw0](https://user-images.githubusercontent.com/92653266/202518760-1528f85d-3f67-488e-ae59-2b0ae25f8ccb.png)

---

We could try pinging one of the [graphql schema](https://github.com/octokit/graphql-schema) maintainers, maybe he knows.
@gr2m

---


UPDATE:

[About scalars](https://docs.github.com/en/graphql/reference/scalars#about-scalars)
[Scalars](https://graphql.github.io/graphql-spec/June2018/#sec-Scalars) are primitive values: Int, Float, String, Boolean, or ID.
> When calling the GraphQL API, you must specify nested subfields until you return only scalars.

**Reply by chilyanin** [reply]
very clear explanation
thank you

**Reply by LangLangBart** [reply]
PS:
1. The validation rule can be found here:  [graphql/graphql-js/.../validation/rules/ScalarLeafsRule](https://github.com/graphql/graphql-js/blob/main/src/validation/rules/ScalarLeafsRule.ts#L40)

> Field "${fieldName}" of type "${typeStr}" must have a selection of subfields. Did you mean "${fieldName} { ... }"?

2. More information on validation: [graphql/graphql-spec/../Validation.md](https://github.com/graphql/graphql-spec/blob/main/spec/Section%205%20--%20Validation.md#validation)

**Reply by LangLangBart** [reply]
PPS: I guess I missed this part while reading the documents. I'm just including it here for completeness until the next person stumbles across this counterintuitive way to write a mutation. [GitHub Docs - Forming calls with GraphQL - About mutations](https://docs.github.com/en/graphql/guides/forming-calls-with-graphql#about-mutations)

> About mutations - To form a mutation, you must specify three things:

> Mutation name. The type of modification you want to perform.
> Input object. The data you want to send to the server, composed of input fields. Pass it as an argument to the mutation name.
> Payload object. The data you want to return from the server, composed of return fields. Pass it as the body of the mutation name.
